### PR TITLE
Support experiment view

### PR DESCRIPTION
**Pull request by SparkSnail**
NNICTL supports start an experiment in a read-only mode.

### CODE REVIEW

File: src/nni_manager/rest_server/restHandler.ts [hunk 1]
```diff
@@ -168,13 +169,20 @@ class NNIRestHandler {
...
                 });
-            } else {
+            } else if (experimentMode === 'resume'){
                 this.nniManager.resumeExperiment().then(() => {
                     res.send();
                 }).catch((err: Error) => {
                     // Resume experiment is a step of initialization, so any exception thrown is a fatal
                     this.handle_error(err, res);
                 });
+            } else {
```
Comment (chicm-ms) [inline src/nni_manager/rest_server/restHandler.ts]: It's better to use else if (experimentMode === 'view'){

File: src/nni_manager/rest_server/restHandler.ts [hunk 2]
```diff
@@ -168,13 +169,20 @@ class NNIRestHandler {
...
+                this.nniManager.viewExperiment().then(() => {
+                    res.send();
+                }).catch((err: Error) => {
+                    // Resume experiment is a step of initialization, so any exception thrown is a fatal
```
Comment (chicm-ms) [inline src/nni_manager/rest_server/restHandler.ts]: This comments need to be updated.

**Comment by QuanluZhang** [comment]
please update the doc

File: src/nni_manager/rest_server/restHandler.ts (hunk 1 above)
Comment (SparkSnail) [reply src/nni_manager/rest_server/restHandler.ts]: updated

File: src/nni_manager/rest_server/restHandler.ts (hunk 2 above)
Comment (SparkSnail) [reply src/nni_manager/rest_server/restHandler.ts]: updated.

**Comment by SparkSnail** [comment]
> please update the doc

Updated doc.

File: src/nni_manager/common/utils.ts [hunk 3]
```diff
@@ -276,7 +276,7 @@ function prepareUnitTest(): void {
     Container.snapshot(TrainingService);
     Container.snapshot(Manager);
 
-    setExperimentStartupInfo(true, 'unittest', 8080);
+    setExperimentStartupInfo('new', 'unittest', 8080);
```
Comment (QuanluZhang) [inline src/nni_manager/common/utils.ts]: better to define macro for it, and use capital

File: src/nni_manager/core/nnimanager.ts [hunk 4]
```diff
@@ -180,13 +180,19 @@ class NNIManager implements Manager {
...
+    }
+
     public async resumeExperiment(): Promise<void> {
         this.log.info(`Resuming experiment: ${this.experimentProfile.id}`);
         //Fetch back the experiment profile
         const experimentId: string = getExperimentId();
         this.experimentProfile = await this.dataStore.getExperimentProfile(experimentId);
         const expParams: ExperimentParams = this.experimentProfile.params;
-
+        this.setStatus('VIEWING');
```
Comment (QuanluZhang) [inline src/nni_manager/core/nnimanager.ts]: why set 'VIEWING' when resuming an experiment?

File: src/nni_manager/core/nnimanager.ts [hunk 5]
```diff
@@ -180,13 +180,19 @@ class NNIManager implements Manager {
         return this.experimentProfile.id;
     }
 
+    public async viewExperiment(): Promise<void> {
+        this.log.info(`Viewing experiment: ${this.experimentProfile.id}`);
+        this.experimentProfile = await this.dataStore.getExperimentProfile(getExperimentId());
+        this.setStatus('VIEWING');
```
Comment (QuanluZhang) [inline src/nni_manager/core/nnimanager.ts]: in VIEWING mode, do we need to disable some buttons in webui and disable some functionalities in `nnictl`?

File: tools/nni_cmd/launcher.py [hunk 6]
```diff
@@ -371,8 +371,9 @@ def launch_experiment(args, experiment_config, mode, config_file_name, experimen
             exit(1)
     log_dir = experiment_config['logDir'] if experiment_config.get('logDir') else None
     log_level = experiment_config['logLevel'] if experiment_config.get('logLevel') else None
-    if log_level not in ['trace', 'debug'] and (args.debug or experiment_config.get('debug') is True):
-        log_level = 'debug'
+    if mode != 'view':
```
Comment (QuanluZhang) [inline tools/nni_cmd/launcher.py]: is it necessary to check whether mode is 'view' or not?

File: tools/nni_cmd/launcher.py [hunk 7]
```diff
@@ -406,84 +407,85 @@ def launch_experiment(args, experiment_config, mode, config_file_name, experimen
...
-
-    #set pai config
-    if experiment_config['trainingServicePlatform'] == 'pai':
-        print_normal('Setting pai config...')
-        config_result, err_msg = set_pai_config(experiment_config, args.port, config_file_name)
-        if config_result:
-            print_normal('Successfully set pai config!')
-        else:
-            if err_msg:
+    if mode != 'view':
```
Comment (QuanluZhang) [inline tools/nni_cmd/launcher.py]: could refactor this part.

File: src/nni_manager/common/utils.ts (hunk 3 above)
Comment (SparkSnail) [reply src/nni_manager/common/utils.ts]: Fixed, use a constant to describe experiment startup info.

File: src/nni_manager/core/nnimanager.ts (hunk 4 above)
Comment (SparkSnail) [reply src/nni_manager/core/nnimanager.ts]: A mis-operation, removed.

File: tools/nni_cmd/launcher.py (hunk 6 above)
Comment (SparkSnail) [reply tools/nni_cmd/launcher.py]: Add debug mode in experiment view.

File: tools/nni_cmd/launcher.py (hunk 7 above)
Comment (SparkSnail) [reply tools/nni_cmd/launcher.py]: refactored.

File: src/nni_manager/rest_server/restHandler.ts [hunk 8]
```diff
@@ -139,11 +139,18 @@ class NNIRestHandler {
...
-            });
+            let experimentMode: string = getExperimentMode();
+            if( experimentMode !== ExperimentStartUpMode.VIEW) {
+                this.nniManager.updateExperimentProfile(req.body, req.query.update_type).then(() => {
+                    res.send();
+                }).catch((err: Error) => {
+                    this.handle_error(err, res);
+                });
+            } else {
+                let message = `Could not update experiment in view mode!`;
```
Comment (QuanluZhang) [inline src/nni_manager/rest_server/restHandler.ts]: @lvybriage could webui correctly handle this message?

File: src/nni_manager/rest_server/restHandler.ts [hunk 9]
```diff
@@ -193,18 +208,26 @@ class NNIRestHandler {
...
             '/experiment/cluster-metadata', expressJoi(ValidationSchemas.SETCLUSTERMETADATA),
             async (req: Request, res: Response) => {
-            // tslint:disable-next-line:no-any
-            const metadata: any = req.body;
-            const keys: string[] = Object.keys(metadata);
-            try {
-                for (const key of keys) {
-                    await this.nniManager.setClusterMetadata(key, JSON.stringify(metadata[key]));
+            let experimentMode: string = getExperimentMode();
+            if(experimentMode !== ExperimentStartUpMode.VIEW) {
```
Comment (QuanluZhang) [inline src/nni_manager/rest_server/restHandler.ts]: white space

File: src/nni_manager/rest_server/restHandler.ts [hunk 10]
```diff
@@ -193,18 +208,26 @@ class NNIRestHandler {
...
+                } catch (err) {
+                    // setClusterMetata is a step of initialization, so any exception thrown is a fatal
+                    this.handle_error(NNIError.FromError(err), res, true);
                 }
-                res.send();
-            } catch (err) {
-                // setClusterMetata is a step of initialization, so any exception thrown is a fatal
-                this.handle_error(NNIError.FromError(err), res, true);
+            } else {
+                let message = `Could not set cluster-metadata in view mode!`;
```
Comment (QuanluZhang) [inline src/nni_manager/rest_server/restHandler.ts]: does nnictl handle this error message?

File: src/nni_manager/rest_server/restHandler.ts [hunk 11]
```diff
@@ -234,21 +257,36 @@ class NNIRestHandler {
...
-            });
+            let experimentMode: string = getExperimentMode();
+            if(experimentMode !== ExperimentStartUpMode.VIEW) {
+                this.nniManager.addCustomizedTrialJob(JSON.stringify(req.body)).then(() => {
+                    res.send();
+                }).catch((err: Error) => {
+                    this.handle_error(err, res);
+                });
+            } else {
+                let message = `Could not add customized trial in view mode!`;
```
Comment (QuanluZhang) [inline src/nni_manager/rest_server/restHandler.ts]: same

File: src/nni_manager/rest_server/restHandler.ts [hunk 12]
```diff
@@ -234,21 +257,36 @@ class NNIRestHandler {
...
-            });
+            let experimentMode: string = getExperimentMode();
+            if(experimentMode !== ExperimentStartUpMode.VIEW) {
+                this.nniManager.cancelTrialJobByUser(req.params.id).then(() => {
+                    res.send();
+                }).catch((err: Error) => {
+                    this.handle_error(err, res);
+                });
+            } else {
+                let message = `Could not delete trial job in view mode!`;
```
Comment (QuanluZhang) [inline src/nni_manager/rest_server/restHandler.ts]: @lvybriage

File: src/nni_manager/rest_server/restHandler.ts (hunk 10 above)
Comment (QuanluZhang) [reply src/nni_manager/rest_server/restHandler.ts]: yeah, nnictl handles this error message

File: src/nni_manager/rest_server/restHandler.ts (hunk 8 above)
Comment (SparkSnail) [reply src/nni_manager/rest_server/restHandler.ts]: Fixed, send 400 code back.

File: src/nni_manager/rest_server/restHandler.ts (hunk 9 above)
Comment (SparkSnail) [reply src/nni_manager/rest_server/restHandler.ts]: removed.

**Comment by rabbit008** [comment]
#1414

File: src/nni_manager/common/experimentStartupInfo.ts [hunk 13]
```diff
@@ -84,6 +88,12 @@ class ExperimentStartupInfo {
         return this.logLevel;
     }
 
+    public getReadonly(): boolean {
```
Comment (QuanluZhang) [inline src/nni_manager/common/experimentStartupInfo.ts]: could be `isReadonly`

File: src/nni_manager/common/log.ts [hunk 14]
```diff
@@ -134,14 +134,21 @@ class Logger {
...
+     * @param param the params to be written
+     */
     private log(level: string, param: any[]): void {
-        const buffer: WritableStreamBuffer = new WritableStreamBuffer();
-        buffer.write(`[${(new Date()).toLocaleString()}] ${level} `);
-        buffer.write(format(param));
-        buffer.write('\n');
-        buffer.end();
-        this.bufferSerialEmitter.feed(buffer.getContents());
+        if (!getReadonly()) {
```
Comment (QuanluZhang) [inline src/nni_manager/common/log.ts]: we can discuss how to write log in read only mode

File: src/nni_manager/core/nnimanager.ts [hunk 15]
```diff
@@ -180,73 +194,78 @@ class NNIManager implements Manager {
...
-            trialData = trialData.concat(<Object[]>JSON.parse(oneImportedData));
-        }
-        this.trialDataForTuner = JSON.stringify(trialData);
-
-        if (this.experimentProfile.execDuration < this.experimentProfile.params.maxExecDuration &&
-            this.currSubmittedTrialNum < this.experimentProfile.params.maxTrialNum &&
-            this.experimentProfile.endTime) {
-            delete this.experimentProfile.endTime;
+        this.readonly = readonly;
+        if (!readonly) {
```
Comment (QuanluZhang) [inline src/nni_manager/core/nnimanager.ts]: if (readonly) return

File: src/nni_manager/common/experimentStartupInfo.ts (hunk 13 above)
Comment (SparkSnail) [reply src/nni_manager/common/experimentStartupInfo.ts]: fixed.

File: src/nni_manager/core/nnimanager.ts (hunk 15 above)
Comment (SparkSnail) [reply src/nni_manager/core/nnimanager.ts]: fixed.

File: src/nni_manager/common/utils.ts [hunk 16]
```diff
@@ -33,7 +33,7 @@ import * as util from 'util';
 
 import { Database, DataStore } from './datastore';
 import { ExperimentStartupInfo, getExperimentId, getExperimentStartupInfo, setExperimentStartupInfo } from './experimentStartupInfo';
-import { Manager } from './manager';
+import { Manager, ExperimentStartUpMode } from './manager';
```
Comment (chicm-ms) [inline src/nni_manager/common/utils.ts]: Is this import necessary?  it seems not used.

File: src/nni_manager/core/nniDataStore.ts [hunk 17]
```diff
@@ -29,7 +29,7 @@ import { Database, DataStore, MetricData, MetricDataRecord, MetricType,
 import { NNIError } from '../common/errors';
 import { getExperimentId, isNewExperiment } from '../common/experimentStartupInfo';
 import { getLogger, Logger } from '../common/log';
-import { ExperimentProfile,  TrialJobStatistics } from '../common/manager';
+import { ExperimentProfile,  TrialJobStatistics, ExperimentStartUpMode } from '../common/manager';
```
Comment (chicm-ms) [inline src/nni_manager/core/nniDataStore.ts]: It seems the new import not used.

File: src/nni_manager/common/log.ts [hunk 18]
```diff
@@ -134,14 +137,21 @@ class Logger {
...
-        buffer.write('\n');
-        buffer.end();
-        this.bufferSerialEmitter.feed(buffer.getContents());
+        if (!this.readonly) {
+            const buffer: WritableStreamBuffer = new WritableStreamBuffer();
+            buffer.write(`[${(new Date()).toLocaleString()}] ${level} `);
+            buffer.write(format(param));
+            buffer.write('\n');
+            buffer.end();
+            this.bufferSerialEmitter.feed(buffer.getContents());
```
Comment (chicm-ms) [inline src/nni_manager/common/log.ts]: Has the dispatcher log been handled?

File: src/nni_manager/common/utils.ts (hunk 16 above)
Comment (SparkSnail) [reply src/nni_manager/common/utils.ts]: removed.

File: src/nni_manager/core/nniDataStore.ts (hunk 17 above)
Comment (SparkSnail) [reply src/nni_manager/core/nniDataStore.ts]: removed.

File: src/nni_manager/common/log.ts (hunk 18 above)
Comment (SparkSnail) [reply src/nni_manager/common/log.ts]: In readonly mode, the dispatcher process will not started.

**Pull request merged**